gridded_etl_tools/utils/extractor.py

```diff
@@ -0,0 +1,363 @@
+
+        Example
+        -------
+        my_extractor = FTPExtractor(my_dataset_manager)
+        with my_extractor("ftp.cdc.noaa.gov") as extractor:
+            # get source files
+
+        Returns
+        -------
+        FTPConnection
+            this object
+
+        Raises
+        ------
+        ValueError
+            If FTPExtractor.
+        """
+        if not hasattr(self, "host"):
+            raise ValueError("FTPExtractor must have a host parameter to open connection")
+        else:
+            self.dm.info(f"Opening a connection to {self.host}")
+            self.ftp = ftplib.FTP(self.host)
+            self.ftp.login()
+            return self
+
+    def __exit__(self, *exception):
+        """
+        Close the connection with the FTP server at this object's given source. This will be called automatically when
+        exiting the connection context manager.
+
+        Parameters
+        ----------
+        *exception
+            Exception information passed automatically by Python
+        """
+        self.ftp.close()
+        self.dm.info(f"Closed connection to {self.host}")
+
+    def __call__(self, host: str):
+        """
+        When called like a function, set the host parameter. See `FTPExtractor.__enter__`.
+
+        Parameters
+        ----------
+        host
+            Address to connect to for source data
+        """
+        self.host = host
+        return self
+
+    @property
+    def cwd(self) -> pathlib.Path:
+        """
+        Returns
+        -------
+        pathlib.Path
+            The object's working directory on the FTP server
+
+        Raises
+        ------
+        RuntimeError
+            If the FTP connection is not open yet
+        """
+        try:
+            return pathlib.Path(self.ftp.pwd())
+        except ftplib.error_perm:
+            raise RuntimeError(
+                "FTP connection must be opened from a context manager before getting the working directory."
+            )
+
+    @cwd.setter
+    def cwd(self, path: pathlib.Path):
+        """
+        Change working directory on the FTP server to the given path. The connection must already be opened using
+        `FTPExtractor.__enter__`.
+
+        Parameters
+        ----------
+        path
+            Directory path to change to
+
+        Raises
+        ------
+        RuntimeError
+            If there is an error during changing directories. This can be caused by the directory not existing, or
+            the connection being closed.
+        """
+        if not self.ftp.nlst(str(path)):
+            raise RuntimeError(f'Could not find path "{path}" on FTP server.')
+        try:
+            self.ftp.cwd(str(path))
+        except ftplib.error_perm:
+            raise RuntimeError("Error changing directory. Is the FTP connection open?")
+
+    def request(self, source: pathlib.Path, destination: pathlib.Path = pathlib.Path()) -> bool:
```

Comment:
Would it be worth it to log a warning saying that this won't actually result in parallel downloading of files? I could see a user being confused as to why the FTPExtractor is downloading one file at a time